cache.go

```
@@ -1,9 +1,110 @@
-package cache_lib_go
+package cache_lib
 
-import "log"
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"log"
+	"time"
 
-func Cache() error {
-	log.Print("Does cache, placeholder for real code")
+	"github.com/go-redis/redis/v8"
+)
+
+type Cache[Data any] interface {
+	RememberBlocking(ctx context.Context, fn LongFunc[Data], key string, ttl time.Duration) (*Data, error)
+}
+
+type cache[Data any] struct {
+	client *redis.Client
+}
+
+type LongFunc[Data any] func(ctx context.Context) (*Data, error)
+
+func NewCache[Data any](client *redis.Client) Cache[Data] {
+	return &cache[Data]{
+		client: client,
+	}
+}
+
+func (c *cache[Data]) getCachedData(ctx context.Context, key string) *Data {
+	cachedData, _ := c.client.Get(ctx, key+"_success").Result()
+
+	if cachedData != "" {
+		var marshaledData Data
+		_ = json.Unmarshal([]byte(cachedData), &marshaledData)
```

It would be better to handle this error, so that it doesn't silently return a default struct if someone made a non-backward-compatible change on the schema.